### PR TITLE
New segmented sort algorithm

### PR DESCRIPTION
**Pull request by senior-zero**
This PR includes a new segmented sort facility. Few approaches to this problem exist. 

### Embed segment number into keys
This approach provides an elegant solution to the load-balancing issue but can lead to slowdowns. It also can't be applicable if the number of bites representing segments number exceeds a maximal number of bytes used by keys. 

### Modified merge sort approach
This idea is implemented in modern GPU.  I've used this approach as a reference for comparison with the new segmented sort algorithm. As I show below, this approach can be outperformed in most cases.

### Kernel specialisation
The idea behind this approach is to partition input segments into size groups. Specialised kernels can further process each size group. The LRB approach initially discussed in [the issue](https://github.com/NVIDIA/cub/issues/224) falls into this category. It also represents the approach that the new segmented sort algorithm relies on.

 
I'm going to briefly describe the genesis of the new segmented sort algorithm to justify some design decisions. 

To minimise the number of kernel specialisations, I've benchmarked different approaches to small (under a few hundred items) segment sorting. I've benchmarked single-thread even-odd sorting, bitonic warp sorting and newly added warp merge sort. The warp-scope merge-sort approach demonstrated some advantages: it can sort bigger segments and outperforms other methods (in the majority of cases). 

Warp-scope merge sort is included in this PR as a separate facility. It's possible to partition architectural warp into multiple virtual ones to sort multiple segments simultaneously. The warp-scope merge sort duplicated a significant part of the previously introduced block-scope merge sort, so I extracted the merge-sort strategy into a separate facility. Both warp and block sort share this strategy. 

Here's the speedup of warp-scope merge sort over warp-bitonic sort:
image

And the speedup of warp-scope merge sort over single-thread odd-even sort:
image

In the figures above I vary segment sizes and segments number. 

To further increase the performance of warp-scope merge sort I needed to load and store data efficiently. I needed warp-scope load, store and exchange facilities. These facilities are also provided in this PR. 

Using a proper sorting algorithm was not enough. Initially, I've assigned a CUDA thread block to a segment. Although this approach demonstrated speedup over the existing one, it led to inefficient resource utilisation because most threads were idle.  Nonetheless, a kernel like this is used as a fallback solution when there are not enough segments. If idle threads don't block other CTAs from execution, there's no reason to spend time on segments partitioning. The fallback kernel helped to eliminate cases when the partitioning stage led to the overall slowdown of the new algorithm.

image

Initially, I implemented a single kernel for all size groups. Depending on the CTA number, I allocated a different number of threads per segment. That is, if the segment size exceeded a few thousand items, I've used slow block-scope radix sort. If the data was about a thousand items and fit into shared memory, I've used in-shared-memory block-scope radix sort. In all these cases, the whole CTA was assigned to a single segment. If the CTA number exceeded the number of large segments, I've partitioned CTA into multiple warp groups, each processing a separate segment. It happened that the large-segment branch limited the occupancy of small-segment one. So I separated this kernel into two. One kernel processes large segments and contains two branches - in-shared-memory sort and slow block-scope sort. Another kernel processes medium and small segments. 

To overlap small/large segments processing, I've used concurrent kernels. This PR contains a single-stream implementation, though. The multi-stream API is out of the scope of this PR and might be introduced later.

### Segments partitioning
The LRB approach discussed in the initial issue balances segments in a non-deterministic way. In some cases, it led to slowdowns because consecutive sub-warps might not end up processing consecutive segments. Here's the speedup of the LRB approach compared to the partitioning based approach.

image

I've tried applying LRB only to the large segments group. This approach also leads to controversial performance results. In rare cases, when an extremely large segment is located in the tail of the large segments group, LRB leads to performance improvements. Otherwise, there are enough segments to overlap its processing, and the LRB stage leads to slowdowns. Therefore, we decided to opt-in LRB. The API with pre-partitioned segments is going to be implemented later. Here's the speedup of the version where I apply LRB to the large segment group.

image

Instead of LRB, I've implemented a three-way partition facility, which is included in this PR. It leads to deterministic partitioning of segments and might be used outside of the segmented sort. The three-way partitioning stage is faster than the LRB stage. 

### Temporary storage layout
The temporary storage layout of the new segmented sort is quite complex. To simplify the code and make it safer, I've introduced temporary storage layout wrappers, which can be found in this PR. 

### Performance
To benchmark the new algorithm on real-world data, I've converted sparse matrices collection to segment sizes. Segments number is equal to the number of rows in a sparse matrix, while segment size is equal to the number of non-zero values in this row. Here's the speedup of the new algorithm compared to the `cub::DeviceSegmentedRadixSort` on A100 GPU while sorting pairs of `std::uint32_t` and `std::uint64_t`.

image

|  speedup   | new algorithm | modernGPU |
| ----------- | ----------- | ----------- |
| min      | 1.11       | 0.16 |
| median  | 61.24        | 7.93 |
| max  | 5314.44  | 2219.73 |

The speedup depends on the segments number:
image

### CODE REVIEW

**Comment by senior-zero** [comment]
run tests

**Comment by zasdfgbnm** [comment]
Hi, do we know how much this PR will benefit well-balanced segmented sort with a small number of segments? For example, 1 -256 segment, 120000 elements in each segment.

**Comment by senior-zero** [comment]
> Hi, do we know how much this PR will benefit well-balanced segmented sort with a small number of segments? For example, 1 -256 segment, 120000 elements in each segment.

Hello, @zasdfgbnm! There should be no significant speedup in this case. The modern GPU version of the segmented sort might perform better in the case of very few large segments. But its speedup doesn't scale well in the range of segments that you've mentioned. 

image

**Comment by zasdfgbnm** [comment]
@senior-zero Noted, thanks for the info!

File: cub/device/dispatch/dispatch_three_way_partition.cuh [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+#pragma once
+
+#include <stdio.h>
+#include <iterator>
+
+#include "dispatch_scan.cuh"
```
Comment (allisonvacanti) [inline cub/device/dispatch/dispatch_three_way_partition.cuh:33]: Prefer include paths from the project root instead of relative paths in new code (e.g.

```
#include <cub/device/dispatch/dispatch_scan>
```

File: cub/agent/agent_segmented_radix_sort.cuh [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+CUB_NAMESPACE_BEGIN
+
+
+template <
+  bool      IS_DESCENDING,                  ///< Whether or not the sorted-order is high-to-low
+  typename  SegmentedPolicyT,               ///< Chained tuning policy
+  typename  KeyT,                           ///< Key type
+  typename  ValueT,                         ///< Value type
+  typename  OffsetT>                        ///< Signed integer type for global offsets
+struct AgentSegmentedRadixSort
```
[the remark this answers is not in the document]
Comment (senior-zero) [reply cub/agent/agent_segmented_radix_sort.cuh:46]: It's implementing `DeviceSegmentedRadixSort`.  `DeviceSegmentedSort` relies on this class for large/medium segments only. To call it `AgentSegmentedSort ` I'd also need to integrate `AgentSubWarpSort` into it. `AgentSubWarpSort` is used in separate kernels, I wouldn't like to mix those two.

File: test/test_device_three_way_partition.cu [hunk 3]
```diff
@@ -0,0 +1,607 @@
...
+    TestDependent<T>(num_items);
+    TestDependent<T>(num_items + 31);
+  }
+}
+
+template <typename T>
+void Test()
+{
+  TestEmpty<T>();
+  TestDependent<T>();
```
Comment (allisonvacanti) [inline test/test_device_three_way_partition.cu:591]: I'm not sure what is meant by "Dependent" in this context.

File: test/test_device_three_way_partition.cu [hunk 4]
```diff
@@ -0,0 +1,607 @@
...
+template <typename T>
+void TestDependent(int num_items)
+{
+  TestStability<T>(num_items);
+  TestEmptyFirstPart<T>(num_items);
+  TestEmptySecondPart<T>(num_items);
+  TestEmptyUnselectedPart<T>(num_items);
+  TestUnselectedOnly<T>(num_items);
+  TestReverseIterator<T>(num_items);
+  TestSingleOutput<T>(num_items);
```
Comment (allisonvacanti) [inline test/test_device_three_way_partition.cu:574]: Nice edge case coverage 👍

File: test/test_temporary_storage_layout.cu [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+#include "test_util.h"
+
+#include <memory>
+
+template <int Items>
+std::size_t GetTemporaryStorageSize(std::size_t (&sizes)[Items])
+{
+  void *pointers[Items]{};
+  std::size_t temp_storage_bytes{};
+  cub::AliasTemporaries(nullptr, temp_storage_bytes, pointers, sizes);
```
Comment (allisonvacanti) [inline test/test_temporary_storage_layout.cu:11]: Check that this returns `cudaSuccess` just to be safe. (no such code path exists when passed `nullptr`, but I'm pedantic :D)

File: cub/agent/agent_segmented_radix_sort.cuh (hunk 2 above)
[the remark this answers is not in the document]
Comment (allisonvacanti) [reply cub/agent/agent_segmented_radix_sort.cuh:46]: > It's implementing `DeviceSegmentedRadixSort`

`DeviceSegmentRadixSort` is not using this agent:

https://github.com/senior-zero/cub/blob/main-feature/github/seg_sort/cub/device/dispatch/dispatch_radix_sort.cuh#L397-L404

[the remark this answers is not in the document]
Comment (allisonvacanti) [reply cub/agent/agent_segmented_radix_sort.cuh:46]: That sounds reasonable, thanks for the context!

To avoid confusion in the meantime, could you file a bug proposing that we switch to in-kernel loops with the details from your initial investigation, and leave a comment on the class's implementation that refers to this bug?

File: cub/device/device_partition.cuh [hunk 6]
```diff
@@ -256,6 +256,178 @@ struct DevicePartition
...
+     * @param[in,out] temp_storage_bytes Reference to size in bytes of
+     *                                   @p d_temp_storage allocation
+     * @param[in] d_in Pointer to the input sequence of data items
+     * @param[out] d_first_part_out Pointer to the output sequence of data items
+     *                              selected by @p select_first_part_op
+     * @param[out] d_second_part_out Pointer to the output sequence of data items
+     *                               selected by @p select_second_part_op
+     * @param[out] d_unselected_out Pointer to the output sequence of unselected
+     *                              data items
+     * @param[out] d_num_selected_out Pointer to the output array with two elements,
```
[the remark this answers is not in the document]
Comment (senior-zero) [reply cub/device/device_partition.cuh:379]: Won't it be inconsistent with the single partition above? It writes only the number of selected items.

File: cub/device/dispatch/dispatch_segmented_sort.cuh [hunk 7]
```diff
@@ -0,0 +1,1434 @@
...
+
+    constexpr int radix_bits = LargeSegmentPolicyT::RADIX_BITS;
+
+    cudaError error = cudaSuccess;
+
+    do
+    {
+      if (num_items == 0 || num_segments == 0)
+      {
+        temp_storage_bytes = 0;
```
[the remark this answers is not in the document]
Comment (senior-zero) [reply cub/device/dispatch/dispatch_segmented_sort.cuh:968]: > Does this behave as expected, and do we have tests for this conditions?

We have `TestZeroSegments` and `TestEmptySegments` for this purpose. 

> IIRC, cudaMalloc(0) returns nullptr, which will throw off the d_temp_storage pointer check later on in the algorithm. 

This check doesn't depend on the `d_temp_storage`. Since the algorithm exits at this point, there won't be any later checks.

File: cub/device/dispatch/dispatch_three_way_partition.cuh [hunk 8]
```diff
@@ -0,0 +1,508 @@
...
+    SelectIfKernelPtrT          three_way_partition_kernel,
+    KernelConfig                three_way_partition_config)
+  {
+    cudaError error = cudaSuccess;
+
+    do
+    {
+      // Get device ordinal
+      int device_ordinal;
+      if (CubDebug(error = cudaGetDevice(&device_ordinal)))
```
[the remark this answers is not in the document]
Comment (senior-zero) [reply cub/device/dispatch/dispatch_three_way_partition.cuh:266]: cub::CurrentDevice doesn't return cudaError_t. I think we should propagate the error code. Is there a way to do so with `cub::CurrentDevice`?

File: test/test_device_segmented_sort.cu [hunk 9]
```diff
@@ -0,0 +1,1542 @@
...
+using namespace cub;
+
+
+constexpr static int MAX_ITERATIONS = 4;
+
+
+class SizeGroupDescription
+{
+public:
+  SizeGroupDescription(const int segments,
```
[the remark this answers is not in the document]
Comment (senior-zero) [reply test/test_device_segmented_sort.cu:51]: I don't like aggregate initialization, [here is](https://godbolt.org/z/5cf5KhW7b) why.

File: cub/warp/warp_store.cuh [hunk 10]
```diff
@@ -0,0 +1,536 @@
...
+    {
+      StoreDirectStriped<LOGICAL_WARP_THREADS>(linear_tid,
+                                               block_itr,
+                                               items,
+                                               valid_items);
+    }
+  };
+
+
+  template <int DUMMY>
```
[the remark this answers is not in the document]
Comment (senior-zero) [reply cub/warp/warp_store.cuh:282]: Explicit specializations are not allowed in the class scope. Here's [an example](https://godbolt.org/z/9r49KE7cn).

File: cub/warp/warp_merge_sort.cuh [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+ * @tparam KeyT
+ *   Key type
+ *
+ * @tparam ITEMS_PER_THREAD
+ *   The number of items per thread
+ *
+ * @tparam LOGICAL_WARP_THREADS
+ *   <b>[optional]</b> The number of threads per "logical" warp (may be less
+ *   than the number of hardware warp threads). Default is the warp size of the
+ *   targeted CUDA compute-capability (e.g., 32 threads for SM86).
```
[the remark this answers is not in the document]
Comment (senior-zero) [reply cub/warp/warp_merge_sort.cuh:52]: Yes, I'm adding the "Must be a power of two" sentence here. I didn't want to duplicate this requirement here since we already have a static assert. But writing it explicitly seems to be reasonable. Otherwise, this information won't be present in the web version of the documentation.

File: cub/warp/warp_exchange.cuh [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+   *
+   * @tparam OffsetT <b>[inferred]</b> Signed integer type for local offsets
+   *
+   * @param[in,out] items Items to exchange
+   * @param[in] ranks Corresponding scatter ranks
+   */
+  template <typename OffsetT>
+  __device__ __forceinline__ void
+  ScatterToStriped(InputT items[ITEMS_PER_THREAD],
+                   OffsetT ranks[ITEMS_PER_THREAD])
```
[the remark this answers is not in the document]
Comment (senior-zero) [reply cub/warp/warp_exchange.cuh:329]: I am afraid this is not a new API, so I can't change it. [Here ](https://github.com/NVIDIA/cub/blob/05bb8419353ed9d3724a42ab69d7bad7950c2d98/cub/block/block_exchange.cuh#L1214) you can find this particular method before my commits.

**Comment by senior-zero** [comment]
> It's about six times slower to compile the new version. The binary size is about 5 times larger.
I've removed all the policies except one for both `cub::DeviceSegmentedRadixSort` and `cub::DeviceSegmentedSort` to have a more honest comparison. After this change, the compilation time is only 2x slower and the binary size is only 1.5x. So the main source of the compilation time issues is the tuning for the wider GPU architectures set.

**Comment by senior-zero** [comment]
@allisonvacanti, thank you for your comments! I've addressed them in the "Fix review notes" commit and also rebased this branch. I'm looking forward to your opinion on these changes.
